lightning-background-prober/src/lib.rs

```diff
@@ -0,0 +1,820 @@
+								last_probe.update_path(path.clone());
+							}
+							return path;
+						},
+						Err(_) => panic!("No route found"), //to be handled later
+					}
+				},
+				
+				Err(_) => panic!("Invalid target_pubkey"),
+			}
+		}
+	}
+}
+
+
+//Will convert Path into Box<[bitcoin::secp256k1::PublicKey]>
+fn path_parser (path: Path) -> Box<[bitcoin::secp256k1::PublicKey]> {
+	let vec_pubkeys = {
+		path.
+		hops.
+		into_iter().
+		map(|hop| hop.pubkey).
+		collect::<Vec<bitcoin::secp256k1::PublicKey>>()
+	};
+	let publickey: Box<[bitcoin::secp256k1::PublicKey]> = vec_pubkeys.into_boxed_slice();
+	return publickey;
+}
+
+
+//capacity = balance(A) + balance(B)
+//liquidity(A) = balance(A) – channel_reserve(A) – pending_HTLCs(A)
+//Will return Sorted Vec of channel_capacity in given Path
+
+// we are halving the channel capacity to get balance of A and B
+// top nodes will try to directional balance, hence channel capcity is closet to real channel liquidity 
+fn sorted_channel_capacity_in_path(
+	network_graph: &ReadOnlyNetworkGraph, 
+	path: Path
+) -> Vec<(u64,Option<u64>)> {
+
+	let mut vec: Vec<(u64,Option<u64>)> = Vec::new();
+	let hops = path.hops;
+		for i in hops {
+			let scid = i.short_channel_id;
+			let capacity_msat = network_graph.channel(scid).unwrap().capacity_sats;
+			match capacity_msat {
+				Some(x) => {
+					let c = (0.5*(x as f64)).floor() as u64;
+					vec.push((scid, Some(c)));
+				}
+				None => vec.push((scid, None)),
+			}
+		}
+
+		vec.sort_by(|a, b| {
+			match (a.1, b.1) {
+				(Some(x), Some(y)) => x.cmp(&y),
+				(None, Some(_)) => std::cmp::Ordering::Greater,
+				(Some(_), None) => std::cmp::Ordering::Less,
+				(None, None) => std::cmp::Ordering::Equal,
+			}
+		});
+
+	return vec;
+}
+
+struct HtlcList {
+    path: Path,
+    max_htlc_list: Vec<(NodeId, u64)>,
+}
+
+impl HtlcList {
+    fn new(path: Path) -> Self {
+        Self {
+            path,
+            max_htlc_list: Vec::new(),
+        }
+    }
+
+    fn sorted_htlc_in_path(&mut self, channel_info: &ChannelInfo) {
+        self.max_htlc_list.clear(); // Clear the list before adding new values
+
+        for i in &self.path.hops {
+            let nodeid = NodeId::from_pubkey(&i.pubkey);
+            let max = channel_info.get_directional_info(0).unwrap().htlc_maximum_msat;
+            self.max_htlc_list.push((nodeid, max));
+        }
+        
+        self.max_htlc_list.sort_by(|(_, a), (_, b)| a.cmp(b)); // ascending ordering of MAX_HTLC
+    }
+}
+
+
+//Value selector for probing (Uses binary search method)
+async fn value_selector (
+	path: Path,
+	nodeid: NodeId,
+	probe_value_map: &ProbeValueMap,
+	max_htlc_list: Vec<(NodeId,u64)>,
+	capacity_list: Vec<(u64,Option<u64>)>,
+) -> u64 {
+	let mut value_map = probe_value_map.value_map.lock().await;
+	let temp = value_map.get_mut(&nodeid);
+	match temp {
+		Some(node_rating) => { 
+			
+			let vec_len = node_rating.probes.len();
+		
+			if vec_len == 0 {
```

Comment:
Why do we need a `Vec` for `probes`? What's the significance of checking its length?

Reply:
It is used to know number of probes send till now, as first probe values depend on `max_htlc_list` and `channel_capacity_list`, I didnt found any good method, without any counter (It might not be updated correctly) or `Vec.len()`

Reply:
> It is used to know number of probes send till now, as first probe values depend on `max_htlc_list` and `channel_capacity_list`,

Could you describe in detail how the `Vec` is used? I'm not entirely convinced it is necessary. Does the number of probes in the `Vec` ever decrease? We can't have a data structure that is constantly growing.

> I didnt found any good method, without any counter (It might not be updated correctly) or `Vec.len()`

I'm not sure what you mean by this.

Reply:
There has to be some counter to know "How many probes has been sent to `target_node`" and `Vec` keeps count of that alone with various info like `probe_value`, as we discussed First probe, don't have any knowledge of path, hence it uses `min(max_htlc_list[0].1, capacity_list[0].1.unwrap())`

If `Vec.len() == 0` hence no probe has been sent
else if `Vec.len() == 1` hence 1 probe has been sent

And for every newly network graph graph, which is currently 24hrs, complete `NodeRating` (Local Probe persister) will be dropped and new instance will be created

Reply:
> There has to be some counter to know "How many probes has been sent to `target_node`" and `Vec` keeps count of that alonge with various info like `probe_value`, as we discussed First probe, don't have any knowledge of path, hence it uses `min(max_htlc_list[0].1, capacity_list[0].1.unwrap())`
> 
> If `Vec.len() == 0` hence no probe has been sent else if `Vec.len() == 1` hence 1 probe has been sent
> 

Why not an `Option`? When would we need to know about more than one probe?

> And for every newly network graph graph, which is currently 24hrs, complete `NodeRating` (Local Probe persister) will be dropped and new instance will be created

That's still needlessly using memory that isn't needed during the 24 hours, no?

Reply:
Got your point, it only need 2 probes persisted, one is 'last successful probe' and second is 'last probe', will make the required changes